File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.java
class Solution {
    public void reversee(StringBuilder str, int start, int end)
    {
        while(start < end)
        {
        char temp = str.charAt(start);
        str.setCharAt(start, str.charAt(end));
        str.setCharAt(end , temp);
        start++;
        end--;
        }
    }
    public String reverseWords(String s) {
       int n = s.length();
       int i = 0 ;
       int j = n -1;

       while(i < n && s.charAt(i)==' ')i++; 
       while(j >= 0 && s.charAt(j)==' ')j--;
        

        
StringBuilder str = new StringBuilder();

       while(i <= j) 
       {
        if(s.charAt(i) != ' ')
        {
            str.append(s.charAt(i));
        }
        
          else if(str.charAt(str.length() -1) != ' ')
           {
             str.append(' ');
           }
    
          i++;
       }

      str.reverse();
     
     int end = 0;
    for(int start = 0 ; start < str.length() ;start++)
    {
        if(str.charAt(start) == ' ')
        {
            reversee(str,end,start-1);
            end = start+1;
        }
    }

    reversee(str,end, str.length()-1);

    return str.toString();
    }
}

File: 0151-reverse-words-in-a-string/0151-reverse-words-in-a-string.cpp
#include <iostream>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
 
    void reverse(int start, int end, string &s) {
        while (start < end) {
            swap(s[start], s[end]);
            start++;
            end--;
        }
    }

    string reverseWords(string s) {
        int n=s.size();
        int i=0;
        int j=0;

    while(i<n && s[i]==' '){
        i++;
    }

while(i<n){
    while(i<n && s[i]!=' '){
        s[j++]=s[i++];
        
       
     }
     while(i<n && s[i]==' '){
        i++;
     }

     if(i<n){
        s[j++]=' ';
     }
}

s.resize(j);

reverse(0,j-1,s);
int start=0;
for(int i=0;i<=j;i++){
    if(i==j||s[i]==' '){
        reverse(start,i-1,s);
        start=i+1;
      }
      
}

  return s;      
    }
};

